kratos/containers/pointer_vector.h

```diff
@@ -267,11 +267,21 @@ class PointerVector final
         mData.swap(rOther.mData);
     }
 
-    void push_back(TPointerType x)
+    void push_back(const TPointerType& x)
```

Comment:
Don't tell me that current push_back was doing copies?, that explains a lot... I think this is not a minor change anymore...

Reply:
A Ok, it is a pointer, in that case is not that bad

Reply:
We should not use emplace_back here because we don't  want to store the original (smart) pointer here without adding to the reference counter. And if it is a normal pointer then it make no difference passing by const reference or value

Reply:
this is `push_back`; `emplace_back` is another member function.

For now, let's forget about raw pointers because the compiler will optimize the differences between copy/move operations away in that case. When it comes to smart pointers, there are 2 situations in userland:
1) The user creates a new managed object from scratch for the sole purpose of inserting it into `PointerVector`.
2) The user inserts an existing smart pointer into `PointerVector`.

Currently, there's only one function to do both jobs:
https://github.com/KratosMultiphysics/Kratos/blob/b228896684903e6f2aefcae73e508f483d6f33fc/kratos/containers/pointer_vector.h#L270-L273

## Current Behaviour

### 1
```cpp
// userland
PointerVector<SmartPointer> pointer_vector;
SmartPointer pointer; // <== a new manged object is created
pointer_vector.push_back(pointer); // <== first copy happens here: 'pointer' is passed by value (moving would make no difference because push_back takes 'pointer' by value anyway)

// pointer_vector.h
void push_back(TPointerType x) 
{ 
    mData.push_back(x); // <== second copy happens here: 'x' is passed by reference and then copied at least once within std::vector
} 
```

*Ignoring the internals of `mData` that is an `std::vector`, here's the tally: there was 1 default construction and 2 copy constructions, resulting in incrementing the reference count a whopping 3 times (at least, because we're ignoring the internals of `std::vector`) and decrementing it twice.*

The user can help the compiler by constructing the managed object in place:
```cpp
// userland
pointer_vector.push_back(SmartPointer()); // <== non-mandatory copy elision
```
*This effectively eliminates the first copy, because the compiler is **allowed** to elide it and construct the managed object directly in the argument of `push_back`. However, the copy within `push_back` cannot be elided and so the tally is: 1 default construction, 1 copy construction ==> 2 reference count increments and 1 decrement.*

### 2
```cpp
// userland
PointerVector<SmartPointer> pointer_vector;
SmartPointer pointer; // <== a new manged object is created
pointer_vector.push_back(pointer); // <== first copy

// pointer_vector.h
void push_back(TPointerType x) 
{ 
    mData.push_back(x); // <== second copy within std::vector
} 
```
*Same thing happens as the case without copy elision. If we assume that `pointer` was already constructed and will survive the container's lifetime, we had 2 refcount increments and 1 decrement.*

## New Behaviour

### 1
```cpp
// userland
PointerVector<SmartPointer> pointer_vector;
SmartPointer pointer; // a new managed object is created
pointer_vector.push_back(pointer); // <== 'pointer' is now passed by reference so the refcount is untouched

// pointer_vector.h
void push_back(const TPointerType& x) 
{ 
    mData.push_back(x); // <== 'x' is passed by reference again, but is then copied within std::vector
} 
```

*Tally: 1 default construction  and 1 copy construction => 2 refcount increments and 1 decrement.*

Constructing in place has now completely different behaviour, because there's a move overload for `push_back`:
```cpp
// userland
PointerVector<SmartPointer> pointer_vector;
pointer_vector.push_back(SmartPointer()); <== a new managed object is created and the move constructor is elided

// pointer_vector.h
void push_back(TPointerType&& x) 
{ 
    mData.push_back(std::move(x)); // <== 'x' is moved, and then move constructed within `std::vector`
} 
```
*Tally: 1 Default construction, 3 moves => **1 refcount increment**.*

Even better, we can now use `emplace_back`:
```cpp
// userland
PointerVector<SmartPointer> pointer_vector;
pointer_vector.emplace_back(); // <== nothing happens here

// pointer_vector.h
template <class ...TArgs>
void emplace_back(TArgs&&... rArgs) 
{ 
    mData.emplace_back(std::forward<TArgs>(rArgs)...); // <== nothing happens here either (rArgs is empty), but the managed object is constructed somewhere down the line
} 
```

*Tally: 1 default construction => **1 refcount increment**.*

### 2
```cpp
// userland
PointerVector<SmartPointer> pointer_vector;
SmartPointer pointer; // a new managed object is created
pointer_vector.push_back(pointer); <== 'pointer' is now passed by reference so the refcount is untouched

// pointer_vector.h
void push_back(const TPointerType& x) 
{ 
    mData.push_back(x); // <== 'x' is passed by reference again, but is then copied within std::vector
} 
```
*If we assume that `pointer` was already constructed and will survive the container's lifetime, we had **1 refcount increment**.*

Reply:
Thanks @matekelemen  for your comprehensive answer!!

Some comments:
1. At first I missed understood you and thought you want to use emplace back inside the push back.
2. This code was written far before move semantic exist in C++
3. In modern C++, to my understanding, if you want to take ownership of the shared pointer you would take the argument by value (to explicitly announce that you are sharing ownership) and user should move it if he wants. (Of course this is arguable)

Saying that I am open to the changes in this branch